doc/developer/design/20231117_copy_to_s3.md

```diff
@@ -0,0 +1,304 @@
+have arrangements spilling to disk in the future, we'll probably get this for free.
+
+Consolidation on read from Persist is not going to be available in this timeline. But once
+it's available we should be able to update this to use that painlessly. In the meantime we should be
+careful that we don't introduce any new uses of memory in the implementation for COPY to S3.
+
+Also, since backpressure is not feasible right off the bat when reading from materialized view as
+I had assumed, there's no reason to limit the scope to only support materialized views/tables and
+we can support arbitrary SELECT queries as well.
+
+### Writing to S3
+While writing to S3, we'll need to split the data into multiple files so that each file is
+under the configured MAX FILE SIZE. Multiple timely workers can do writes in paralle, where each
+worker would do the [multi part uploads](https://docs.aws.amazon.com/AmazonS3/latest/userguide/mpuoverview.html)
+to S3.
+* Each worker starts a multi part upload with a `CreateMultipartUpload` API call to start
+writing data to a S3 file in the given path.
+* The worker should keep a counter, say `bytes_to_write` to keep track of the aggregate bytes
+it will write.
+* The worker should also keep some state to track the number of parts in the upload
+and the corresponding etags returned from AWS for the uploaded parts.
+* There should be a configurable value `max_multi_part_bytes_in_memory` which can default to 100MB.
+This is the max value of the data which the worker will keep in memory before uploading
+a part to the multi-part upload.
+* The worker receives data to be written to S3, it checks if the size of the data exceeds
+configured `max_multi_part_bytes_in_memory`. If so, then the data should be split to keep
+within the limit. The worker then adds the size of the data to `bytes_to_write` to see if
+it will exceed MAX FILE SIZE.
+* If less than MAX FILE SIZE, the worker assigns an incremental part number starting from 1,
+makes a `UploadPart` call and keeps the returned `ETag` along with the part number in memory.
+* If more than MAX FILE SIZE, the worker makes a `CompleteMultipartUpload` call to finish
+the upload of a single file and provides the part number and etags in the request.
+* Then the worker creates a new multi part request to write to a new S3, resets the
+`bytes_to_write`, clears the part number with etag mapping in memory and uploads the part for
+the new multi part call.
+
+#### S3 file names
+With multiple workers writing data to S3 in parallel, the file names can follow a pattern
+like `<prefix>/part-<worker_idx>-<0001>.csv`, `<prefix>/part-<worker_idx>-<0001>.csv`.
+
+## Rollout and Testing
+We should put COPY TO S3 behind a feature flag.
+
+#### Testing after code is merged
+- Switch on the feature flag for the staging environment and do an end to end copy to s3
+flow in staging.
+- Also, try out a scenario where we cancel the copy midway
```

Comment:
We should also add a benchmark (not sure if a feature benchmark or a scalability benchmark here is more appropriate) that copies out a material amount of data and measures how long it takes.